File: src/MyChess/ChessMove.cs
using System;

namespace MyChess;

public class ChessMove : IComparable
{
    public PiecePlayer Player { get; private set; }

    public PieceRank Rank { get; private set; }

    public ChessBoardLocation From { get; private set; }

    public ChessBoardLocation To { get; private set; }

    public ChessSpecialMove SpecialMove { get; private set; }

    public ChessMove(PieceRank rank, PiecePlayer player, int horizontalLocationFrom, int verticalLocationFrom, int horizontalLocationTo, int verticalLocationTo)
    {
        Rank = rank;
        Player = player;
        From = new ChessBoardLocation(horizontalLocationFrom, verticalLocationFrom);
        To = new ChessBoardLocation(horizontalLocationTo, verticalLocationTo);
        SpecialMove = ChessSpecialMove.None;
    }

    public ChessMove(PieceRank rank, PiecePlayer player, int horizontalLocationFrom, int verticalLocationFrom, int horizontalLocationTo, int verticalLocationTo, ChessSpecialMove specialMove)
    {
        Rank = rank;
        Player = player;
        From = new ChessBoardLocation(horizontalLocationFrom, verticalLocationFrom);
        To = new ChessBoardLocation(horizontalLocationTo, verticalLocationTo);
        SpecialMove = specialMove;
    }

    // Must match 'int ChessMove.CompareTo(object obj)' implicitly implemented member 'int IComparable.CompareTo(object? obj)'
#nullable disable
    public int CompareTo(object obj)
    {
        if (obj == null)
        {
            return 1;
        }

        if (obj is ChessMove otherMove)
        {
            int compare = this.From.CompareTo(this.From);
            if (compare == 0)
            {
                compare = this.To.CompareTo(otherMove.To);
                if (compare == 0)
                {
                    compare = this.Player.CompareTo(otherMove.Player);
                    if (compare == 0)
                    {
                        return this.Rank.CompareTo(otherMove.Rank);
                    }
                }
            }

            return compare;
        }
        else
        {
            throw new ArgumentException("Object is not a ChessMove");
        }
    }
#nullable enable

    public override string ToString()
    {
        char horizontalFrom = (char)((int)'A' + this.From.Column);
        char verticalFrom = (char)((int)'1' + (7 - this.From.Row));
        char horizontalTo = (char)((int)'A' + this.To.Column);
        char verticalTo = (char)((int)'1' + (7 - this.To.Row));

        return string.Format("{0}{1}{2}{3}", horizontalFrom, verticalFrom, horizontalTo, verticalTo);
    }
}

File: src/MyChessReact/src/game/ChessMove.ts
import { ChessBoardLocation } from "./ChessBoardLocation";
import { ChessPiece } from "./ChessPiece";
import { ChessSpecialMove } from "./ChessSpecialMove";
import { ChessPlayer } from "./ChessPlayer";

export class ChessMove {
    piece: ChessPiece;
    player: ChessPlayer;

    from: ChessBoardLocation;
    to: ChessBoardLocation;

    specialMove: ChessSpecialMove;

    comment: string;

    time: string

    constructor(piece: ChessPiece, player: ChessPlayer, horizontalLocationFrom: number, verticalLocationFrom: number, horizontalLocationTo: number, verticalLocationTo: number, specialMove?: ChessSpecialMove, comment?: string, time?: string) {
        this.piece = piece;
        this.player = player;
        this.from = new ChessBoardLocation(horizontalLocationFrom, verticalLocationFrom);
        this.to = new ChessBoardLocation(horizontalLocationTo, verticalLocationTo);
        this.specialMove = specialMove != null ? specialMove : ChessSpecialMove.None;
        this.comment = comment != null ? comment : "";
        this.time = time != null ? time : "";
    }

    public compareTo(otherMove: ChessMove): number {
        let compare: number = this.from.compareTo(this.from);
        if (compare === 0) {
            compare = this.to.compareTo(otherMove.to);
            if (compare === 0) {
                if (this.player === otherMove.player) {
                    return 0;
                }
            }
        }

        return compare;
    }

    public getMoveString(): string {
        let horizontalFrom: string = String.fromCharCode(this.from.horizontalLocation + "A".charCodeAt(0));
        let verticalFrom: string = String.fromCharCode("8".charCodeAt(0) - this.from.verticalLocation);
        let horizontalTo: string = String.fromCharCode(this.to.horizontalLocation + "A".charCodeAt(0));
        let verticalTo: string = String.fromCharCode("8".charCodeAt(0) - this.to.verticalLocation);
        return horizontalFrom + verticalFrom + horizontalTo + verticalTo;
    }
}
